server/applicationset/applicationset_test.go

```diff
@@ -557,6 +557,56 @@ func TestUpdateAppSet(t *testing.T) {
 	})
 }
 
+func TestUpsertAppSet(t *testing.T) {
+	name := "test"
+	ns := "external-namespace"
+	appSet := newTestAppSet(func(appset *appsv1.ApplicationSet) {
+		appset.Annotations = map[string]string{
+			"annotation-key1": "annotation-value1",
+			"annotation-key2": "annotation-value2",
+		}
+		appset.Labels = map[string]string{
+			"label-key1": "label-value1",
+			"label-key2": "label-value2",
+		}
+		appset.Name = name
+		appset.Namespace = ns
+		appset.Finalizers = []string{"finalizer"}
+	})
+
+	updatedAppSet := newTestAppSet(func(appset *appsv1.ApplicationSet) {
+		appset.Annotations = map[string]string{
+			"annotation-key1": "annotation-value1-updated",
+		}
+		appset.Labels = map[string]string{
+			"label-key1": "label-value1-updated",
+		}
+		appset.Name = name
+		appset.Namespace = ns
+		appset.Finalizers = []string{"finalizer-updated"}
+	})
+
+	t.Run("Upsert", func(t *testing.T) {
+		appServer := newTestAppSetServer(t, appSet)
+
+		updated, err := appServer.Create(t.Context(), &applicationset.ApplicationSetCreateRequest{
+			Applicationset: updatedAppSet,
+			Upsert:         true,
+		})
+
+		require.NoError(t, err)
+		assert.Equal(t, map[string]string{
```

Comment:
how are we able to get the updated the annotations here? Secondly, what's the use of `appSet` created at [Line **563](https://github.com/argoproj/argo-cd/pull/23860/files#diff-1b078e5493d7d523ef5e4e0ab51e198b4f82eb44475a54b0e37951ebb9fce010R563)?**

Reply:
> Secondly, what's the use of appSet created at Line 563?

This is created to ensure that the call to `Create` is updating/upserting an existing app set and not creating a new one. I.e. triggering the logic [here](https://github.com/argoproj/argo-cd/blob/3e6f11e08e709bb92b933106f8ca97e0ac6b3343/server/applicationset/applicationset.go#L232-L260) since the function [returns early](https://github.com/argoproj/argo-cd/blob/3e6f11e08e709bb92b933106f8ca97e0ac6b3343/server/applicationset/applicationset.go#L223-L227) if a new app set is created.